forem/forem
# Fix JS error when window.Forem.runtime is not available yet

## What type of PR is this? (check all applicable)


## Description

There's an outstanding issue that happens in the FE of end users that relates to `window.Forem` and `window.Forem.runtime` when they are not defined. This seems to happen when `base.jsx` executes after the asset pipeline initializers, causing those functions to be `undefined`.

This error appears in Honeybadger across a few different reports ([example 1](https://app.honeybadger.io/projects/67192/faults/89477500), [example 2](https://app.honeybadger.io/projects/67192/faults/89477551/01GQJPY4TBR4NB9D9TTDN97D15?q=window.forem), [example 3](https://app.honeybadger.io/projects/67192/faults/92779714/01GQHCTGF2VF3JZN5GRNV2A7DK?q=window.forem#notice-trace)). All of those backtraces reveal an attempt to the use `window.Forem`/`window.Forem.runtime` and it being `undefined`.

The reason appears to be that the webpack based bundle entrypoint `base.jsx` hasn't executed the code that initializes these functions yet ([see here](https://github.com/forem/forem/blob/main/app/javascript/packs/base.jsx#L20)). This PR attempts to avoid this by checking if this is available, and if not wait until it is available (200ms at a time).

## Related Tickets & Documents

Perhaps some flaky tests are related to this problem too, since JS execution halts when an error is encountered and this has a waterfall effect on other areas of the JS codebase. This is hard to track though, so not linking any specific issues with this PR for now.

## QA Instructions, Screenshots, Recordings

1. Boot local server
2. Visit `/pod` and attempt to play a podcast
   - In case you can't see Podcasts locally there's a chance you don't have cache enabled locally
   - Run `bundle exec rails dev:cache` and then `Podcasts::Feed.new(Podcast.last).get_episodes(limit: 10)` to fetch episodes
3. Podcast playback should work

### UI accessibility concerns?

N/A

## Added/updated tests?

- [x] No, and this is why: This is an area of the codebase that isn't covered by tests due to many intricacies related to video/podcast playback

## [Forem core team only] How will this change be communicated?


## [optional] Are there any post deployment tasks we need to perform?

Keep an eye out for Honeybadger errors coming in related to `window.Forem`/`window.Forem.runtime` being `undefined`

## [optional] What gif best describes this PR or how it makes you feel?

![not yet](https://media.giphy.com/media/YQAuKJ7wf68qBHPw6Y/giphy.gif)
